Add WKWebsiteDataStore API to create persistent store

szewai opened the pull request:
#### 74f1364427e0cd5a5bf3d6c866a0c19762cf14fd
<pre>
Add WKWebsiteDataStore API to create persistent store
<a href="https://bugs.webkit.org/show_bug.cgi?id=253590">https://bugs.webkit.org/show_bug.cgi?id=253590</a>
rdar://101314896

Reviewed by Geoffrey Garen.

Add API based on existing SPI.

* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore identifier]):
(+[WKWebsiteDataStore dataStoreForIdentifier:]):
(+[WKWebsiteDataStore removeDataStoreForIdentifier:completionHandler:]):
(+[WKWebsiteDataStore fetchAllDataStoreIdentifiers:]):
(+[WKWebsiteDataStore _removeDataStoreWithIdentifier:completionHandler:]):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::dataStoreForIdentifier):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.h:
(WebKit::WebsiteDataStoreConfiguration::create):

Canonical link: <a href="https://commits.webkit.org/261443@main">https://commits.webkit.org/261443@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/884184a2907a849a1723940091593438907225ed

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/111658 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/20794 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/269 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/87/builds/3492 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120396 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22149 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/11874 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/3153 "layout-tests (failure)") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117424 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/16447 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99593 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/104608 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/98399 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/170 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/45376 "") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13273 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/169 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/87806 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13763 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/9619 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19209 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52167 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7965 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15753 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.h:
```diff
@@ -83,6 +83,30 @@ WK_CLASS_AVAILABLE(macos(10.11), ios(9.0))
 @property (nonatomic, readonly) WKHTTPCookieStore *httpCookieStore WK_API_AVAILABLE(macos(10.13), ios(11.0));
 
+/*! @abstract Get identifier for a data store.
+ @discussion Default and non-persistent data store do not have identifier.
```
data store => data stores
have identifier => have an identifier

When we say "do not have", do we mean this property returns nil? If so, let's say nil.

geoffreygaren commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.h:
```diff
@@ -83,6 +83,30 @@ WK_CLASS_AVAILABLE(macos(10.11), ios(9.0))
+
+/*! @abstract Get a persistent data store.
+ @param identifier An identifier that is used to uniquely identify the data store.
+ @discussion If data store with specified identifier does not exists yet, it will be created. It returns nil if data store cannot be created.
```
If data store => If a data store
exists yet => exist yet
It returns => Returns

Can we specify the conditions that would prevent a data store from being created?

szewai replied:
Yes, if identifier is invalid.

geoffreygaren commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.h:
```diff
@@ -83,6 +83,30 @@ WK_CLASS_AVAILABLE(macos(10.11), ios(9.0))
+
+/*! @abstract Delete a persistent data store.
+ @param identifier An identifier that is used to uniquely identify the data store.
+ @param completionHandler A block to invoke with optional error when operation completes.
```
when operation => when the operation

geoffreygaren commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.h:
```diff
@@ -83,6 +83,30 @@ WK_CLASS_AVAILABLE(macos(10.11), ios(9.0))
+/*! @abstract Delete a persistent data store.
+ @param identifier An identifier that is used to uniquely identify the data store.
+ @param completionHandler A block to invoke with optional error when operation completes.
+ @discussion This will remove the whole data store so it should be called when the data store is not and will not be used any more. Error is returned if removal fails to complete.
```
Let's state explicitly that the developer must release any WKWebView and/or WKWebsiteDataStore that uses the data store before invoking this function.

geoffreygaren commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.h:
```diff
@@ -83,6 +83,30 @@ WK_CLASS_AVAILABLE(macos(10.11), ios(9.0))
++ (void)removeDataStoreForIdentifier:(NSUUID *)identifier completionHandler:(void(^)(NSError *))completionHandler WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
+
+/*! @abstract Fetch all data store identifiers.
+ @param completionHandler A block to invoke with an array of identifiers when operation completes.
```
when operation => when the operation

geoffreygaren commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.h:
```diff
@@ -83,6 +83,30 @@ WK_CLASS_AVAILABLE(macos(10.11), ios(9.0))
+
+/*! @abstract Fetch all data store identifiers.
+ @param completionHandler A block to invoke with an array of identifiers when operation completes.
+ @discussion Default or non-persistent data store do not have identifier.
```
data store => data stores
have identifier => have an identifier

geoffreygaren reviewed (approved):
r=me

I have some suggestions about wording.

szewai committed:
Add WKWebsiteDataStore API to create persistent store
https://bugs.webkit.org/show_bug.cgi?id=253590
rdar://101314896

Reviewed by Geoffrey Garen.

Add API based on existing SPI.

* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore identifier]):
(+[WKWebsiteDataStore dataStoreForIdentifier:]):
(+[WKWebsiteDataStore removeDataStoreForIdentifier:completionHandler:]):
(+[WKWebsiteDataStore fetchAllDataStoreIdentifiers:]):
(+[WKWebsiteDataStore _removeDataStoreWithIdentifier:completionHandler:]):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::dataStoreForIdentifier):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.h:
(WebKit::WebsiteDataStoreConfiguration::create):

Canonical link: https://commits.webkit.org/261443@main
